File: week6/pset6/credit.py
from cs50 import get_int


def main():
    while True:
        n = get_int("Number: ")
        if n >= 0:  # checks for positive number
            break

    if VALIDATE(n):
        PRINTBRAND(n)  # function to print CC brans
    else:
        print("INVALID")


def VALIDATE(n):
    return CHECKSUM(n)


def CHECKSUM(n):
    sum = 0
    for i in range(len(str(n))):
        if (i % 2 == 0):
            sum += n % 10  # takes the sum of digits not multiplied by 2
        else:
            digit = 2 * (n % 10)  # multiplies every other digit by 2
            sum += digit // 10 + digit % 10  # splits the digits of 2-digit numbers

        n //= 10

    return sum % 10 == 0
    

def PRINTBRAND(n):
    if (n >= 34e13 and n < 35e13) or (n >= 37e13 and n < 38e13):
        print("AMEX")  # for cards starting with 34/37
    elif ((n >= 51e14 and n < 56e14)):
        print("MASTERCARD")  # for cards starting with 51,52,53,54,55
    elif ((n >= 4e12 and n < 5e12) or (n >= 4e15 and n < 5e15)):
        print("VISA")  # for cards starting with 4
    else:
        printf("INVALID")


if __name__ == "__main__":
    main()

File: week1/pset1/credit.c
//includes
#include <stdio.h>
#include <cs50.h>

//functions
bool VALIDATE(long long n);
int LENGTH(long long n);
bool CHECKSUM(long long n);
void PRINTBRAND(long long n);

int main(void)
{
    long long n; // Where n is a variable for credit card number
    
    do
    {
        n = get_long_long("Number: ");
        // Receives input from user
    }
    while (n < 0);
    
    if (VALIDATE(n) == true)
    {
        PRINTBRAND(n);
        // If number is valid, credit card brand is printed
    }
    else
    {
        printf("INVALID\n");
        // For invalid inputs
    }
}

bool VALIDATE(long long n)
{
    int len = LENGTH(n);
    return ((len == 13 || len == 15 || len == 16) && CHECKSUM(n));
    // Checks for correct number of digits and Luhn's Algorithm
}

int LENGTH(long long n)
{
    int len;
    for (len = 0; n != 0; n /= 10, len++);
    return len;
}

bool CHECKSUM(long long n) // Function for Luhn's Algorithm
{
    int sum = 0;
    for (int i = 0; n != 0; i++, n /= 10)
    {
        if (i % 2 == 0)
        {
            sum += n % 10;
        }
        else
        {
            int digit = 2 * (n % 10);
            sum += digit / 10 + digit % 10;
        }
    }
    return (sum % 10) == 0;
}

void PRINTBRAND(long long n) // Print instructions based on starting digits
{
    if ((n >= 34e13 && n < 35e13) || (n >= 37e13 && n < 38e13))
    {
        printf("AMEX\n");
    }
    else if ((n >= 51e14 && n < 56e14))
    {
        printf("MASTERCARD\n");
    }
    else if ((n >= 4e12 && n < 5e12) || (n >= 4e15 && n < 5e15))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
  
